pkg/identity/cidr.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identity
+
+import (
+	"net"
+
+	"github.com/cilium/cilium/pkg/labels/cidr"
+)
+
+// AllocateCIDRIdentities allocates identities for each of the specified CIDR
+// prefixes so they will be available later during policy resolution.
+//
+// On success, the returned slice will have a 1-to-1 correspondence with the
+// CIDR in the provided prefixes slice, eg the identity for prefixes[0] will
+// be held in identities[0].
+// On failure, attempts to clean up after itself and returns the error.
+func AllocateCIDRIdentities(prefixes []*net.IPNet) (res []*Identity, err error) {
+	if len(prefixes) == 0 {
+		return nil, nil
+	}
+
+	log.Debugf("Attempting to allocate identities for %d prefixes", len(prefixes))
+
+	// Allocate identities for each CIDR.
+	res = make([]*Identity, 0, len(prefixes))
+	for _, prefix := range prefixes {
+		if prefix == nil {
+			continue
+		}
+		labels := cidr.GetCIDRLabels(prefix)
+		id, _, err := AllocateIdentity(labels)
+		if err != nil {
+			// If any identity allocation failed, release existing identities
+			// and log the error.
+			if err2 := ReleaseSlice(res); err2 != nil {
+				log.WithError(err2).Error("Could not recover from error during CIDR identity allocation")
+			}
+			res = nil
+			break
+		}
+		res = append(res, id)
+	}
+
+	if err == nil {
+		log.Debugf("Allocated identities for %d prefixes", len(res))
+	} else {
+		log.WithError(err).Warningf("Failed to allocate identities for %d prefixes", len(res))
+	}
+
+	return res, err
+}
+
+// LookupCIDRIdentities finds identities corresponding to each of the specified
+// prefixes. It expects to be able to find all identities.
+// On success, returns a slice of identities with a 1-to-1 correspondence to
+// the specified slice of prefixes, and nil.
+// On error, returns all identities that can be resolved and an error.
+func LookupCIDRIdentities(prefixes []*net.IPNet) (res []*Identity, err error) {
+	res = make([]*Identity, 0, len(prefixes))
+	err = nil
+
+	for _, prefix := range prefixes {
+		labels := cidr.GetCIDRLabels(prefix)
+		id, err2 := identityAllocator.Get(globalIdentity{labels})
+		if err2 == nil {
+			res = append(res, NewIdentity(NumericIdentity(id), labels))
+		} else {
+			err = err2
+			log.Infof("Cannot locate identity for CIDR %s", prefix.String())
+		}
+	}
+
+	return res, err
```

Comment:
It's unfortunate that we only return the most recent error here. Similarly, we don't associate the error with the identity we were trying to get. I'm not sure it matters, though, since the errors here should be network type things, right? not anything specific to an identity?

Reply:
I believe that it shouldn't be anything to do with the identity, I've tested with a variety of different generated sets of labels including ones for prefixes like "::/128".

The assumption here is very much that if there's one problem, there may be many problems, and they're unlikely to differ.

Perhaps when I look at putting a controller in charge here, we could be a bit more thorough in our error reporting.